### ISSUE TITLE
Consider adding class support (defining canister as a class)

### ISSUE THREAD

**Issue by lastmjs** [demergent-labs/azle#24]
- [x] Rewrite all examples with class API at the bottom
- [x] Remove old JS generation stuff
- [ ] attempt to get rid of the CanisterResult type if possible
- [x] Make sure all tests pass
- [x] Rewrite parsing
- [x] Remove old API
- [x] Make sure all tests pass
- [x] Add tsconfig.json to documentation now that we need the experimentalDecorators field
- [x] Make sure stable storage examples use the correct stable storage memory (they will overwrite each other if you are not careful)
- [x] Consider all of these alternative ways to create decorator-like things: https://stackoverflow.com/questions/35949554/invoking-a-function-without-parentheses
- [x] Explore the $ syntax
- [x] Explore combining new with a special type that will allow new Canister(Principal) without requiring the user to use a class to define the external canister, and also create the return type for them automatically...ideally they would define the external canister in the same way as they define their local canister methods

Consider creating support for classes. I would really like to avoid encouraging object oriented design, but canisters are stateful, they have methods, and the class syntax example shown below does map very well.

```typescript
class IC {
    caller: Principal;
    // memory: {
    //     balances: {
    //         [key: AccountIdentifier]: Balance | undefined;
    //     };
    //     extensions: Extension[];
    //     initialized: boolean;
    //     supply: Balance;
    // };

    get memory(): {
        balances: {
            [key: AccountIdentifier]: Balance | undefined;
        };
        extensions: Extension[];
        initialized: boolean;
        supply: Balance;
    } {
        return ic.memory;
    }
}

export class JSONIC extends IC {
    constructor() {
        super();

        this.memory.initialized = true;
        this.memory.balances = {};
        this.memory.supply = 0;
        this.memory.extensions = ['@ext/common'];
    }

    @update
    balance(request: BalanceRequest): BalanceResponse {
        const aid = getUserAID(request.user);
    
        const balance = ic.memory.balances[aid];
    
        if (balance === undefined) {
            return {
                ok: 0
            };
        }
    
        return {
            ok: balance
        };
    }
}
```

**Comment by lastmjs** [comment]
More notes

```typescript
class Canister<Memory, Storage> {
    caller: Principal;
    memory: Memory;
    storage: Storage;
    // balances: {
    //     [key: AccountIdentifier]: Balance | undefined;
    // };
    // extensions: Extension[];
    // initialized: boolean;
    // supply: Balance;
    // memory: {
    //     balances: {
    //         [key: AccountIdentifier]: Balance | undefined;
    //     };
    //     extensions: Extension[];
    //     initialized: boolean;
    //     supply: Balance;
    // };

    // get memory(): {
    //     balances: {
    //         [key: AccountIdentifier]: Balance | undefined;
    //     };
    //     extensions: Extension[];
    //     initialized: boolean;
    //     supply: Balance;
    // } {
    //     return ic.memory;
    // }

    // get initialized()

    // get balances() {
    //     return ic.memory.balances;
    // }
}

type Memory = {
    balances: {
        [key: AccountIdentifier]: Balance | undefined;
    };
    extensions: Extension[];
    initialized: boolean;
    supply: Balance;
};

type Storage = {};

export class JSONIC extends Canister<Memory, Storage> {
    constructor() {
        super();

        this.memory.initialized = true;
        this.memory.balances = {};
        this.memory.supply = 0;
        this.memory.extensions = ['@ext/common'];
    }

    @update
    balance(request: BalanceRequest): BalanceResponse {
        const aid = getUserAID(request.user);
    
        const balance = ic.memory.balances[aid];
    
        if (balance === undefined) {
            return {
                ok: 0
            };
        }
    
        return {
            ok: balance
        };
    }
}
```

**Comment by dansteren** [comment]
Another plus side to using classes is that you can initialize variables more cleanly. For example, in the [whoami example](https://github.com/demergent-labs/azle/tree/main/examples/motoko_examples/whoami), in examples/motoko_examples/whoami/src/whoami.ts we have to initialize state variables with initial values before the `init` function is called. If we were to write canisters as classes you would be able to declare a `constructor` method and initialize state variables in there. This would guarantee that the variables would be set because the constructor is called before everything else.

**Comment by dansteren** [comment]
A big driving factor on this is implementing function guards. They could look pretty crazy if we don't have decorators, so implementing this could be very beneficial.

**Comment by dansteren** [comment]
While attempting to convert our examples into the class-based structure we found two major DX setbacks:

- Splitting a canister into multiple files can be _**very**_ cumbersome and adds a ton of boilerplate
- Passing class functions around requires using arrow functions with lexical scoping or adding `.bind(this)`. Otherwise the `this` is incorrect inside the callbacks.

With that in mind, we believe the DX for classes (although better in some instances), is largely inferior to our current approach. As such we are experimenting with other ways to get the benefits of decorators and may pursue those options.

TLDR; we will not proceed with this class-based approach at this time.

**Comment by lastmjs** [comment]
Complications that classes would bring:

- [ ] Confusion over context (this), .bind and similar problems
- [ ] Confusion over where to store global state, on the class or globally
- [ ] Splitting classes across multiple files
- [ ] Using the default export which seems to have become a code smell for web devs
- [ ] Knowing when to use class methods vs creating functions outside of the class.

**Comment by dansteren** [comment]
Superseded by https://github.com/demergent-labs/azle/pull/891.
